File: rental/src/test/java/io/axoniq/demo/bikerental/rental/paymentsaga/PaymentSagaTest.java
package io.axoniq.demo.bikerental.rental.paymentsaga;

import io.axoniq.demo.bikerental.coreapi.payment.PaymentConfirmedEvent;
import io.axoniq.demo.bikerental.coreapi.payment.PaymentRejectedEvent;
import io.axoniq.demo.bikerental.coreapi.payment.PreparePaymentCommand;
import io.axoniq.demo.bikerental.coreapi.rental.ApproveRequestCommand;
import io.axoniq.demo.bikerental.coreapi.rental.BikeRequestedEvent;
import io.axoniq.demo.bikerental.coreapi.rental.RejectRequestCommand;
import io.axoniq.demo.bikerental.rental.support.CommandDispatcher;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.concurrent.CompletableFuture;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Unit tests for the in-memory {@link PaymentSaga}, driven by feeding it events and verifying the
 * commands it dispatches through a mocked {@link CommandDispatcher}.
 */
class PaymentSagaTest {

    private static final String BIKE_ID = "bikeId";
    private static final String RENTER = "rider";
    private static final String REFERENCE = "rentalReference";

    private CommandDispatcher commandDispatcher;
    private PaymentSaga saga;

    @BeforeEach
    void setUp() {
        commandDispatcher = mock(CommandDispatcher.class);
        when(commandDispatcher.send(any())).thenReturn(CompletableFuture.completedFuture(null));
        saga = new PaymentSaga(commandDispatcher);
    }

    @Test
    void bikeRequestPreparesPayment() {
        saga.on(new BikeRequestedEvent(BIKE_ID, RENTER, REFERENCE));

        verify(commandDispatcher).send(new PreparePaymentCommand(10, REFERENCE));
    }

    @Test
    void confirmedPaymentApprovesRequest() {
        saga.on(new BikeRequestedEvent(BIKE_ID, RENTER, REFERENCE));

        saga.on(new PaymentConfirmedEvent("paymentId", REFERENCE));

        verify(commandDispatcher).send(new ApproveRequestCommand(BIKE_ID, RENTER));
    }

    @Test
    void rejectedPaymentRejectsRequest() {
        saga.on(new BikeRequestedEvent(BIKE_ID, RENTER, REFERENCE));

        saga.on(new PaymentRejectedEvent("paymentId", REFERENCE));

        verify(commandDispatcher).send(new RejectRequestCommand(BIKE_ID, RENTER));
    }
}

File: rental/src/paymentsaga/payment-saga.test.ts
import assert from 'node:assert/strict';
import { beforeEach, test } from 'node:test';
import {
  ApproveRequestCommand,
  BikeRequestedEvent,
  type CommandDispatcher,
  PaymentConfirmedEvent,
  PaymentRejectedEvent,
  PreparePaymentCommand,
  RejectRequestCommand,
} from '@bikerental/core-api';
import { PaymentSaga } from './payment-saga.js';

const BIKE_ID = 'bikeId';
const RENTER = 'rider';
const REFERENCE = 'rentalReference';

let sent: object[];
let saga: PaymentSaga;

function recordingDispatcher(): CommandDispatcher {
  return {
    send: (payload: object) => {
      sent.push(payload);
      return Promise.resolve();
    },
  } as unknown as CommandDispatcher;
}

beforeEach(() => {
  sent = [];
  saga = new PaymentSaga(recordingDispatcher());
});

test('bikeRequestPreparesPayment', () => {
  saga.on(new BikeRequestedEvent(BIKE_ID, RENTER, REFERENCE));

  assert.ok(sent.some((c) => deepEquals(c, new PreparePaymentCommand(10, REFERENCE))));
});

test('confirmedPaymentApprovesRequest', () => {
  saga.on(new BikeRequestedEvent(BIKE_ID, RENTER, REFERENCE));
  saga.on(new PaymentConfirmedEvent('paymentId', REFERENCE));

  assert.ok(sent.some((c) => deepEquals(c, new ApproveRequestCommand(BIKE_ID, RENTER))));
});

test('rejectedPaymentRejectsRequest', () => {
  saga.on(new BikeRequestedEvent(BIKE_ID, RENTER, REFERENCE));
  saga.on(new PaymentRejectedEvent('paymentId', REFERENCE));

  assert.ok(sent.some((c) => deepEquals(c, new RejectRequestCommand(BIKE_ID, RENTER))));
});

function deepEquals(a: object, b: object): boolean {
  return (
    Object.getPrototypeOf(a) === Object.getPrototypeOf(b) &&
    JSON.stringify(a) === JSON.stringify(b)
  );
}
